Firestore/core/src/model/base_path.h

```diff
@@ -174,6 +184,31 @@ class BasePath {
 
  private:
   SegmentsT segments_;
+
+  static util::ComparisonResult compareSegments(const std::string& lhs,
+                                                const std::string& rhs) {
+    bool isLhsNumeric = isNumericId(lhs);
+    bool isRhsNumeric = isNumericId(rhs);
+
+    if (isLhsNumeric && !isRhsNumeric) {
+      return util::ComparisonResult::Ascending;
+    } else if (!isLhsNumeric && isRhsNumeric) {
+      return util::ComparisonResult::Descending;
+    } else if (isLhsNumeric && isRhsNumeric) {
+      return util::Compare(extractNumericId(lhs), extractNumericId(rhs));
+    } else {
+      return util::Compare(lhs, rhs);
+    }
+  }
+
+  static bool isNumericId(const std::string& segment) {
+    return segment.substr(0, 4) == "__id" &&
+           segment.substr(segment.size() - 2) == "__";
```

Comment:
`substr` throws and exception if `pos` is larger than the string length ([source](https://en.cppreference.com/w/cpp/string/basic_string/substr)). Also not sure what the behavior would be if segment.size is 1 (and therefore segment.size()-2 will be -1). Please add a unit test for these corner cases.

You may want to do:
```cpp
return
  segment.size() > 6 &&
  segment.substr(0, 4) == "__id" &&
  segment.substr(segment.size() - 2) == "__";
```

Reply:
by adding unit tests I mean testing that the output of `isNumericId` is `false` for (`"__id"` or `"__id__"` or `""`) for example.

Reply:
The test cases include documentIds like "12", "7", "A", "Aa", whose length is less than 6 and IsNumericId returns false.

Reply:
""  and "__id__" are not allowed.